crates/ruff_index/src/list.rs

```diff
@@ -1,46 +1,116 @@
+//! Sorted, arena-allocated association lists
+//!
+//! An [_association list_][alist], which is a linked list of key/value pairs. We additionally
+//! guarantee that the elements of an association list are sorted (by their keys), and that they do
+//! not contain any entries with duplicate keys.
+//!
+//! Association lists have fallen out of favor in recent decades, since you often need operations
+//! that are inefficient on them. In particular, looking up a random element by index is O(n), just
+//! like a linked list; and looking up an element by key is also O(n), since you must do a linear
+//! scan of the list to find the matching element. The typical implementation also suffers from
+//! poor cache locality and high memory allocation overhead, since individual list cells are
+//! typically allocated separately from the heap.
+//!
+//! We solve that last problem by storing the cells of an association list in an [`IndexVec`]
+//! arena. You provide the index type (`I`) that you want to use with this arena. That means that
+//! an individual association list is represented by an `Option<I>`, with `None` representing an
+//! empty list.
```

Comment:
This seems to describe the previous API?

Reply:
Yes, good catch, fixed!